### PR TITLE
Update interface and `README.md`

### PR DESCRIPTION
**Pull request by VictorCMiraldo**
This PR addresses #87 and updates the interface in `Pirouette.Symbolic.Prover.Runner` so we can use it in all our tests, regardless of whether they are tests for the `Ex` or `PlutusIR` language. Here's two examples:

```haskell
parms :: IncorrectnessParams Ex
parms = IncorrectnessParams
          [term| \x:Integer . x + 1 |] -- Term to symbolically evaluate
          [ty| Integer |] -- type of the term above
          -- Conditions to check
          ([term| \(res:Integer) (x:Integer) . 0 < res |]
            :==>: [term| \(res:Integer) (x:Integer) . 0 < x |])

-- replIncorrecntessLogic1 receives no definitions, it's used to check a single term.
replIncorrectnessLogic1 10 params
```

Or, for plutus, taken from the `tests/unit/Language/Pirouette/PlutusIR/SymEvalSpec.hs`:
```haskell
          execFromPIRFile
            "tests/unit/resources/fromPlutusIRSpec-01.pir"
            ( IncorrectnessParams
                { ipTarget = [pir| \(x : Integer) . addone x |],
                  ipTargetType = [pirTy| Integer |],
                  ipCondition =
                    [pir| \(result : Integer) (x : Integer) . 0 < result |]
                      :==>: [pir| \(result : Integer) (x : Integer) . 0 < x |]
                }
            )
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,40 +2,75 @@
...
-study and model-check properties over the Plutus Mealy
-Machine.
+Pirouette is being built as framework for constructing different static
+analisys tools for [System F][systemf] based languages. It has a number of
+transformations (monomorphization, defunctionalization, prenex, etc...) defined
+and relies on the `-XQuasiQuotes` extension to be able to easily write
+properties in the target language.
 
-## Plutus Smart Contracts and Transition Systems
+There is one analisys currently built into pirouette, which symbolic evaluates terms up to
```
Comment (lucaspena) [inline README.md:18]:
```suggestion
There is one analysis currently built into pirouette, which symbolically evaluates terms up to
```

File: README.md [hunk 2]
```diff
@@ -2,40 +2,75 @@
...
-the constructors of the `Input` datatype are used to represent the possible state transitions. This means
-that when TLA+ finds a counter-example, we will see the sequence `[i0, ..., iN] :: [Input]` of inputs
-we need to pass to `transition` to witness the failure.
+And we wish to check the incorrectness triple `[x > 0] addone x [result > 0]`, that
+is: if `addone x > 0`, then `x > 0`. We obviously expect a counterexample with `x = 0`.
+We can use pirouette to find that assignment:
+
+```haskell
+parms :: IncorrectnessParams Ex
+parms = IncorrectnessParams
```
Comment (lucaspena) [inline README.md:47]:
```suggestion
params :: IncorrectnessParams Ex
params = IncorrectnessParams
```

File: README.md [hunk 3]
```diff
@@ -2,40 +2,75 @@
...
+[Plutus] is a subset of Haskell used to
+write smart contracts for Cardano blockchain, which utilizes the _Extended UTxO_
+[[1](https://iohk.io/en/research/library/papers/the-extended-utxo-model/),
+[2](https://iohk.io/en/research/library/papers/native-custom-tokens-in-the-extended-utxo-model/)]
+model to represent its accounts. Pirouette has a prototype interface to interacting with
+plutus through `Language.Pirouette.PlutusIR`, but further work is necessary to
+seamlessly interact with validator scripts.
+
+An example of using pirouette to load and check an incorrectness triple over a `.pir` file
+can be found in [here](tests/unit/Language/Pirouette/PlutusIR/SymEvalSpec.hs).
```
Comment (lucaspena) [inline README.md:73]:
```suggestion
can be found [here](tests/unit/Language/Pirouette/PlutusIR/SymEvalSpec.hs).
```

File: README.md [hunk 4]
```diff
@@ -48,56 +83,12 @@ to avoid building GHC when you start the nix shell.
...
-In its current form, `pirouette` itself is still a research prototype and, hence, has its limitations.
-In particular, it has difficulties dealing with arbitrary higher-order PlutusIR code. When pirouette successfully
-outputs a TLA+ spec, some familiarity and experience with TLA+ itself are required to extract useful results
-out of it. The tool does not figure out the specification for you.
-
-The overall approach of using TLA+ also carries the usual limitations
-of a model checker: the process of exhaustively checking the state
+In its current form, `pirouette` itself is still a research prototype
+and, hence, it has plenty of limitations. The overall approach of pirouette's
+symbolic engine is an intersetion of bounded model checking and symbolic executon, hence,
```
Comment (lucaspena) [inline README.md:90]:
```suggestion
symbolic engine is an intersection of bounded model checking and symbolic execution, hence,
```

File: src/Pirouette/Symbolic/Prover/Runner.hs [hunk 5]
```diff
@@ -21,37 +22,69 @@ import qualified Test.Tasty.HUnit as Test
...
-    ipCondition :: AssumeProve lang,
-    ipMaxCstrs :: Int
+  { ipTarget :: Term lang,
+    ipTargetType :: Type lang,
+    ipCondition :: AssumeProve lang
   }
 
+-- | The result type of 'runIncorrectnessLogic', returning the first (if any)
+--  path that is a counterexample to the supplied 'IncorrectnessParams'.
+--  If you would like finer grained control, look at 'execIncorrectnessLogic.
```
Comment (lucaspena) [inline src/Pirouette/Symbolic/Prover/Runner.hs:34]:
```suggestion
--  If you would like finer grained control, look at 'execIncorrectnessLogic'.
```

**Review by lucaspena — COMMENTED** [review]
Looks good to me besides the typos I pointed out

File: README.md [hunk 6]
```diff
@@ -2,40 +2,75 @@
...
+We can use pirouette to find that assignment:
+
+```haskell
+parms :: IncorrectnessParams Ex
+parms = IncorrectnessParams
+          [term| \x:Integer . x + 1 |] -- Term to symbolically evaluate
+          [ty| Integer |] -- type of the term above
+          -- Conditions to check
+          ([term| \(res:Integer) (x:Integer) . 0 < res |]
+            :==>: [term| \(res:Integer) (x:Integer) . 0 < x |])
```
Comment (facundominguez) [inline README.md:52]:
```suggestion
          [term| \x:Integer . x + 1 |] -- A function term to symbolically evaluate
          [ty| Integer |] -- type of the result of the function term above
          -- Conditions to check, they take the result and the inputs of the function term 
          ([term| \(res:Integer) (x:Integer) . 0 < res |]
            :==>: [term| \(res:Integer) (x:Integer) . 0 < x |])
```

File: README.md [hunk 7]
```diff
@@ -48,56 +83,12 @@ to avoid building GHC when you start the nix shell.
...
-outputs a TLA+ spec, some familiarity and experience with TLA+ itself are required to extract useful results
-out of it. The tool does not figure out the specification for you.
-
-The overall approach of using TLA+ also carries the usual limitations
-of a model checker: the process of exhaustively checking the state
+In its current form, `pirouette` itself is still a research prototype
+and, hence, it has plenty of limitations. The overall approach of pirouette's
+symbolic engine is an intersetion of bounded model checking and symbolic executon, hence,
+it also carries the limitations of a model checker: it can be slow
 space can be slow and a model checker is not a substitute for a formal
```
Comment (VictorCMiraldo) [inline README.md:92]:
```suggestion
it also carries the limitations of a model checker: exploration of the state space can be slow
and it is not a substitute for a formal
```

File: README.md [hunk 8]
```diff
@@ -48,56 +83,12 @@ to avoid building GHC when you start the nix shell.
...
 proof of correctness.
```
[the remark this answers is not in the document]
Comment (VictorCMiraldo) [reply README.md:94]: Indeed, it was a copy paste error. I fixed it, thanks!

File: README.md [hunk 9]
```diff
@@ -2,40 +2,75 @@
...
+  x ↦ 0 }
+```
+
+## Plutus Smart Contracts
+
+[Plutus] is a subset of Haskell used to
+write smart contracts for Cardano blockchain, which utilizes the _Extended UTxO_
+[[1](https://iohk.io/en/research/library/papers/the-extended-utxo-model/),
+[2](https://iohk.io/en/research/library/papers/native-custom-tokens-in-the-extended-utxo-model/)]
+model to represent its accounts. Pirouette has a prototype interface to interacting with
```
Comment (0xd34df00d) [inline README.md:68]: I'm not sure the EUTxO part is worth mentioning here in the context of Pirouette.

The folks who know about Plutus likely know this, and the folks who don't probably don't care.

File: README.md [hunk 10]
```diff
@@ -48,57 +83,13 @@ to avoid building GHC when you start the nix shell.
...
-In its current form, `pirouette` itself is still a research prototype and, hence, has its limitations.
-In particular, it has difficulties dealing with arbitrary higher-order PlutusIR code. When pirouette successfully
-outputs a TLA+ spec, some familiarity and experience with TLA+ itself are required to extract useful results
-out of it. The tool does not figure out the specification for you.
-
-The overall approach of using TLA+ also carries the usual limitations
-of a model checker: the process of exhaustively checking the state
-space can be slow and a model checker is not a substitute for a formal
+In its current form, `pirouette` itself is still a research prototype
+and, hence, it has plenty of limitations. The overall approach of pirouette's
```
Comment (0xd34df00d) [inline README.md:89]:
```suggestion
In its current form, `pirouette` itself is still a research prototype,
hence, it has plenty of limitations. The overall approach of pirouette's
```

File: README.md [hunk 11]
```diff
@@ -48,57 +83,13 @@ to avoid building GHC when you start the nix shell.
...
+symbolic engine is an intersection of bounded model checking and symbolic execution, hence,
+it also carries the limitations of a model checker: exploration of the state space can be slow
+and it is not a substitute for a formal
```
Comment (0xd34df00d) [inline README.md:92]: This is indeed very good to mention!

File: src/Pirouette/Symbolic/Prover/Runner.hs [hunk 12]
```diff
@@ -21,37 +22,69 @@ import qualified Test.Tasty.HUnit as Test
...
+    ipTargetType :: Type lang,
+    ipCondition :: AssumeProve lang
   }
 
+-- | The result type of 'runIncorrectnessLogic', returning the first (if any)
+--  path that is a counterexample to the supplied 'IncorrectnessParams'.
+--  If you would like finer grained control, look at 'execIncorrectnessLogic'.
+type IncorrectnessResult lang = Maybe (Path lang (EvaluationWitness lang))
+
+runIncorrectnessLogic1 ::
```
Comment (0xd34df00d) [inline src/Pirouette/Symbolic/Prover/Runner.hs:37]: What's `1` here? At first I thought it's the number of counterexamples to return, but now I'm not sure what's the difference from the un-`1`-ed version.

File: src/Pirouette/Symbolic/Prover/Runner.hs [hunk 13]
```diff
@@ -21,37 +22,69 @@ import qualified Test.Tasty.HUnit as Test
...
+    shouldStop st = sestConstructors st > maxCstrs
+
+    isCounter Path {pathResult = CounterExample _ _, pathStatus = s}
+      | s /= OutOfFuel = True
+    isCounter _ = False
+
+-- | Prepares a 'IncorrectnessParams' into a 'Problem', to be passed to
+--  some worker function. Chances are that you are looking for
+--  'runIncorrectnessLogic' instead, unless you want a specific analysis.
+--  If that's the case, the arguments to worker that you want will
```
Comment (0xd34df00d) [inline src/Pirouette/Symbolic/Prover/Runner.hs:64]: Or `If that's not the case`?

File: tests/unit/Language/Pirouette/PlutusIR/ToTermSpec.hs [hunk 14]
```diff
@@ -7,48 +7,46 @@ import Control.Monad.Reader
...
-      -- Test that expanding one declaration yields the other declaration
-      testCase "expandDefs produces NF terms" $ do
-        decls <- progAct
-        let longN = head $ filter ((== "long") . nameString . snd) $ M.keys decls
-        let shortN = head $ filter ((== "short") . nameString . snd) $ M.keys decls
-        let (DFunction _ long _) = fromJust $ M.lookup longN decls
-        let (DFunction _ short _) = fromJust $ M.lookup shortN decls
-        runReader (expandDefs long) (mocked decls) @?= short
-    ]
+tests = (: []) $
```
Comment (0xd34df00d) [inline tests/unit/Language/Pirouette/PlutusIR/ToTermSpec.hs:19]: Oh, nice, this `: []` is not something I see too often.

**Review by 0xd34df00d — APPROVED** [review]
Minor README stuff mostly.

File: README.md [hunk 15]
```diff
@@ -2,40 +2,75 @@
...
+* [Building, Installing and Hacking](#building-installing-and-hacking)
 * [Contributing](#contributing)
 * [License](#license)
 
-Pirouette is a semi-automatic code extraction tool. It extracts a
-[TLA+](https://lamport.azurewebsites.net/tla/tla.html) specification
-from a [Plutus] Mealy Machine. The extracted TLA+ specification can then be used to
-study and model-check properties over the Plutus Mealy
-Machine.
+Pirouette is being built as framework for constructing different static
```
[the remark this answers is not in the document]
Comment (VictorCMiraldo) [reply README.md:12]: Agree!

File: README.md [hunk 16]
```diff
@@ -2,40 +2,75 @@
...
+analisys tools for [System F][systemf] based languages. It has a number of
```
[the remark this answers is not in the document]
Comment (VictorCMiraldo) [reply README.md:13]: I'll spell check before merging

File: src/Pirouette/Symbolic/Prover/Runner.hs (hunk 12 above)
Comment (VictorCMiraldo) [reply src/Pirouette/Symbolic/Prover/Runner.hs:37]: The one is that it uses a single term, it doesn't take in a set of definitions (`PrtUnorderedDefs`). I'll see if I come up with a better name

File: README.md [hunk 17]
```diff
@@ -2,40 +2,75 @@
...
+transformations (monomorphization, defunctionalization, prenex, etc...) defined
+and relies on the `-XQuasiQuotes` extension to be able to easily write
+properties in the target language.
```
[the remark this answers is not in the document]
Comment (VictorCMiraldo) [reply README.md:16]: Agree, I moved it further down

File: README.md [hunk 18]
```diff
@@ -2,40 +2,75 @@
...
 
-## Plutus Smart Contracts and Transition Systems
+There is one analisys currently built into pirouette, which symbolic evaluates terms up to
+a certain bound while looking for counterexamples based on [incorrectness][incorrectness]
+or Hoare triples.
```
[the remark this answers is not in the document]
Comment (VictorCMiraldo) [reply README.md]: Fair! I expanded it a little in a separate section.

File: src/Pirouette/Symbolic/Prover/Runner.hs (hunk 12 above)
Comment (VictorCMiraldo) [reply src/Pirouette/Symbolic/Prover/Runner.hs]: Renamed to `runIncorrectnessLogicSingl`

**Pull request merged**